### DISCUSSION TITLE
Add cache tags to response headers

### DISCUSSION THREAD

**Discussion by orinokai** [vercel/next.js#57263] [Ideas]
### Goals

Allow for better cache invalidation when using Next.js behind a CDN, by adding a Cache-Tags response header (for pages or route handlers)

### Background

Next.js currently adds the `x-next-cache-tags` header to a static page response, but only when the site is running in minimal mode, i.e. on Vercel.

This header (or a more general header, i.e. Cache-Tags) would be useful to receive when running Next.js behind a CDN (perhaps when in standalone mode) so that the cache can be invalidated by tag alongside Next.js's own cache handling so I am proposing to expose this header more generally.

### Proposal

I'm very happy to contribute a PR and would suggest one of the following options:
1. Use an env var to specify whether a Cache-Tags header should be output
2. Output cache tags when in standalone mode
3. Always output the cache tags

It would be a simple change and the code that should be updated is here: https://github.com/vercel/next.js/blob/a65e5c07a87b27e82f5443afe4b4871f64dc4707/packages/next/src/server/base-server.ts#L2456-L2465

**Comment by diarmuidie** [comment]
Without this it's not possible to implement on demand ISR and revalidation if self hosting behind a caching CDN.

As mentioned in the proposal we would need to expose the already existing tags HTTP header for app router pages, but would also need to generate a new tag header for pages router paths. This could contains the `x-invoked-path` value to use as a unique tag and something that the cache handler is also aware of, that the cache handler can trigger a cache purge on the CDN for pages routes also.

https://github.com/vercel/next.js/blob/15f7418fd6707d541be3a44388404f101bae89be/packages/next/src/server/lib/router-server.ts#L254

**Comment by jespermb** [comment]
Any news on this? It seems like a very important feature for any web framework to allow developers or a third party to be able to extend headers on a route based on the contents of that route.

**Comment by jnm733** [comment]
This feature would be really powerful to work easily with external CDN caches alongside the Next.js internals.

**Comment by RaeesBhatti** [comment]
Relevant PR: https://github.com/vercel/next.js/pull/57264

**Comment by jmotes** [comment]
@orinokai I see that Netlify supports on-demand revalidation now. Can you elaborate on how you were able to work around this limitation?

**Comment by slongdotexe** [comment]
Hey, just jumping in to +1 how useful this could be. 
Recently, AWS CloudFront started supporting invalidation by cache tag: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/invalidation-by-tags.html. So, having a mechanism that lets developers set cache-tag headers could be really powerful for CloudFront and a wide range of CDNs, I would imagine.
